### PR TITLE
New blog article on AWS antipatterns/best practices

### PR DESCRIPTION
**Pull request by crystoll**
Can be reviewed and published, date currently set to today (22.12.)

### CODE REVIEW

File: _posts/2020-12-22-aws-antipatterns.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
+---
+layout: post
+title: AWS Antipatterns
+author: arto
+excerpt: >
+  Cloud platforms offer unprecedented agility, innovation, and speed. Yet I see the same mistakes being repeated over and over again. This blog post is a collection of some things people often get wrong - and some thoughts on how to dodge those mistakes.
```
Comment (denzilferreira) [inline _posts/2020-12-22-aws-antipatterns.md:6]: > get wrong

Comes across as negative :) how about: "This blog post is a collection of some of the pitfalls we unknowingly fall into."

File: _posts/2020-12-22-aws-antipatterns.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+ - Amazon Web Services
+ - Best Practices
+ - Lessons Learned
+ - Antipatterns
+ - Security
+---
+
+![Speed and agility of cloud development can be unsurpassed](/img/aws-antipatterns/aws_rider.png)
+
+Cloud platforms are incredible. For anyone who used to create or operate software running in a traditional on-premises server room, they offer incredible speed in getting ideas to production, and experimentation with new things. However, this speed comes with a cost: Sometimes that cost is money, as AWS bills can sneak up on you. Sometimes the cost may be a bad architecture that gets more expensive as you progress, a badly compromised security that makes you vulnerable for many new attack vectors - or it can be that you are paying for lots of new services just because you can - without getting the benefits you were looking for.
```
Comment (denzilferreira) [inline _posts/2020-12-22-aws-antipatterns.md:18]: How about adding a link to CVE for AWS? 
https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=AWS

File: _posts/2020-12-22-aws-antipatterns.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+In this article, I'm going to go through some of the worst mistakes that might happen when you start working on the AWS platform. These mistakes are based on both personal experience, aka my mistakes, and things I've seen, aka other people's mistakes. I will also try to provide some insights into why we need to pay attention here, and what could we do to avoid the negative impact while benefiting from the newfound speed and agility. I've been working pretty much in cloud and serverless for the last five years, before that I had a lot of on-premises experience. I'm most well versed with AWS architecture and best practices, but most of these thoughts will probably apply also to Azure and Google Cloud Platform projects.
```
Comment (denzilferreira) [inline _posts/2020-12-22-aws-antipatterns.md:20]: > These mistakes are based on both personal experience, aka my mistakes, and things I've seen, aka other people's mistakes

A bit of repetition on aka :) others' mistakes are implied on the "some of the mistakes are personal", right? How about:
Some of the mistakes here are personal 😓 .

File: _posts/2020-12-22-aws-antipatterns.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+## Antipattern #1: Too wide permissions
+
+Normal on-premises systems use firewalls and isolation maintained by network administrators between separate servers and applications, to try and limit accidental access. Or they might not. On the other hand, building and combining services in a cloud environment, there might not be any protection in place, unless you build it yourself. And sadly every AWS tutorial and video training course online tends to do the examples by giving full and wide access for any services that are used - for speed and convenience. And this is exactly the first antipattern we want to shut down.
+
+!['Allow everything' - policy](/img/aws-antipatterns/crazy_wide_policy.png)
+
+The wider concept here is to minimize the **blast radius**. And we do that with the Principle of Least Permission (**POLP**). Blast radius means an area of services and infrastructure that may be affected if a security breach or misconfiguration takes place. In other words, if your lambda has full access to anything in the account, and it gets compromised, damage can be horrendous. Also, if anyone is auditing your solution, this mess will be revealed and require actions to fix.
```
Comment (denzilferreira) [inline _posts/2020-12-22-aws-antipatterns.md:30]: Maybe add link to AWS documentation on permission boundaries?
https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_boundaries.html

File: _posts/2020-12-22-aws-antipatterns.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+There are some tools also to help with this, that can review the templates or policies, and alert if they seem a bit too wide. For example, AWS has services like IAM access advisor, or AWS Trusted Advisor/AWS Organizations can be used to audit this constantly yourself. I've also had success with tools such as CFRipper. There are also plenty of commercial tools if you wish to go that route. You can run these as part of your build, or set up a Cloudformation event-triggered lambda that can stop deployment if the quality gate is not passed. With good tools, it's possible to create exceptions to rules or modify and evolve the rules yourself. But a good tool is also cost-effective and cheap. If you introduce some automation in your development process, this cleanup and tightening are automatically done when it's the cheapest.
```
Comment (denzilferreira) [inline _posts/2020-12-22-aws-antipatterns.md:42]: Maybe add link to this?
https://aws.amazon.com/blogs/security/tighten-s3-permissions-iam-users-and-roles-using-access-history-s3-actions/

File: _posts/2020-12-22-aws-antipatterns.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+Manual work is especially error-prone when something breaks badly and people are under pressure to fix it, or stressed, or hungry, or thinking of their upcoming vacation.. I think you got the point.
+
+![Lots of choices in AWS console - do you remember them all?](/img/aws-antipatterns/manual_work_2.png)
+
+Specific to cloud environments, manual work also implies that developer accounts have and need to have wide access to APIs and services underneath. And it implies that when they make a human mistake, such as dropping a database in the wrong environment, because they had one too few cups of coffee in the morning, the results can be immediate and irreplaceable.
+
+The cure for all of this is to automate, follow the Infrastructure as Code (IaC) principle. By creating any resources by code instead of manual work, we gain a lot of good things:
```
Comment (denzilferreira) [inline _posts/2020-12-22-aws-antipatterns.md:54]: Bold IaC, add the bold style in IaC on the list below?

File: _posts/2020-12-22-aws-antipatterns.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+So: Keep that agility and velocity - but also keep your eyes open. Do also cost-driven development, in addition to all your other viewpoints. Don't do things just because you can, think about costs, benefits, and ROI. That way you can make your development less eventful, but much more peaceful. Fewer surprises equal more happiness, in this case.
+
+
+## Conclusion
+
+For each antipattern, there's a cure. If you build it in your development process, from the very beginning, life will be much better for you. Some of these are impossible to solve on very high and abstract levels, so are best taken care of by empowered DevOps teams that want to carry the responsibility of not only building the services but making sure they run smoothly as well.
```
Comment (denzilferreira) [inline _posts/2020-12-22-aws-antipatterns.md:177]: >there's a cure
 Maybe a mitigation? Somehow "cure" comes across as related to a health condition :-D

File: _posts/2020-12-22-aws-antipatterns.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+- Creating virtual servers, it's possible to choose their type and capacity, and some of them can be very expensive if they are left running. Machine learning projects might often run these servers, but are typically not supposed to run them 24/7, only on demand.
+- Sagemaker developer endpoints can also rack up some bill quite fast, it's easy to create them and use them, but if you forget to shut them down, euros will start accumulating rapidly
+- There are often few ways to model things, and the cost may vary quite a lot. Different patterns depending on your case.
+- If you're not seeing how much things cost, you might be enticed to create a new feature for the product, based on some new service just because you can. Perhaps it's something you wouldn't have even considered before.
+
+This can often be a problem because I've seen that billing is often abstracted, denied, and hidden from the development team. So they can freely
```
Comment (LassiAutioSolita) [inline _posts/2020-12-22-aws-antipatterns.md:166]: "So they can freely", how does this continue?

File: _posts/2020-12-22-aws-antipatterns.md [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+## Conclusion
+
+For each antipattern, there's a cure. If you build it in your development process, from the very beginning, life will be much better for you. Some of these are impossible to solve on very high and abstract levels, so are best taken care of by empowered DevOps teams that want to carry the responsibility of not only building the services but making sure they run smoothly as well.
+
+So remember a lot of acronyms and concepts: Do IaC and POLP, limit the Blast Radius.
+
+I could have picked up a lot more of antipatterns but you gotta start from somewhere, and hopefully, these made you pause for a bit - or smile a happy smile because all this is already taken care of and you are an enlightened developer.
+
+Also remember: You can fake it until you make it. In case you cannot automate everything immediately, including some of these in code reviews is already an improvement. Automation might bring more support for people who would like immediate feedback loops when they create an increment of value, in other words, might be something you want to invest in periodically more and more.
+
```
Comment (denzilferreira) [inline _posts/2020-12-22-aws-antipatterns.md:184]: Maybe add heading: Further reading?

**Review by LassiAutioSolita — CHANGES_REQUESTED** [review]

File: _posts/2020-12-22-aws-antipatterns.md (hunk 1 above)
Comment (crystoll) [reply _posts/2020-12-22-aws-antipatterns.md:6]: I like that idea! Fixing..

File: _posts/2020-12-22-aws-antipatterns.md (hunk 2 above)
Comment (crystoll) [reply _posts/2020-12-22-aws-antipatterns.md:18]: Mmm, I'd rather not, as security is only one of the topics here, and spooking about that list is kinda out of proportion for this article.

File: _posts/2020-12-22-aws-antipatterns.md (hunk 3 above)
Comment (crystoll) [reply _posts/2020-12-22-aws-antipatterns.md:20]: I like this repetition, it's intentional here :)

File: _posts/2020-12-22-aws-antipatterns.md (hunk 4 above)
Comment (crystoll) [reply _posts/2020-12-22-aws-antipatterns.md:30]: Good idea!

File: _posts/2020-12-22-aws-antipatterns.md (hunk 5 above)
Comment (crystoll) [reply _posts/2020-12-22-aws-antipatterns.md:42]: Perhaps not, this is steering in the domain of ops work, I try to keep on general advice level, again we could go deeper and more concrete on any topic but here I would choose not to.

File: _posts/2020-12-22-aws-antipatterns.md (hunk 7 above)
Comment (crystoll) [reply _posts/2020-12-22-aws-antipatterns.md:177]: Yeah, I like that more too.

File: _posts/2020-12-22-aws-antipatterns.md (hunk 9 above)
Comment (crystoll) [reply _posts/2020-12-22-aws-antipatterns.md:184]: Yes, good catch, this part was a bit unfinished. I'll add the heading and the other link discussed above, it goes nicely in here.

File: _posts/2020-12-22-aws-antipatterns.md [hunk 10]
```diff
@@ -0,0 +1,198 @@
+---
+layout: post
+title: AWS Antipatterns
+author: arto
+excerpt: >
+  Cloud platforms offer unprecedented agility, innovation, and speed. Yet I see the same mistakes being repeated over and over again. This blog post is a collection of some things people often get wrong - and some thoughts on how to dodge those mistakes.
+tags:
+ - AWS
+ - Amazon Web Services
```
[the remark this answers is not in the document]
Comment (crystoll) [reply _posts/2020-12-22-aws-antipatterns.md:9]: Yes, good idea.

File: _posts/2020-12-22-aws-antipatterns.md [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+
+## Antipattern #1: Too wide permissions
+
+Normal on-premises systems use firewalls and isolation maintained by network administrators between separate servers and applications, to try and limit accidental access. Or they might not. On the other hand, building and combining services in a cloud environment, there might not be any protection in place, unless you build it yourself. And sadly every AWS tutorial and video training course online tends to do the examples by giving full and wide access for any services that are used - for speed and convenience. And this is exactly the first antipattern we want to shut down.
+
+!['Allow everything' - policy](/img/aws-antipatterns/crazy_wide_policy.png)
```
[the remark this answers is not in the document]
Comment (crystoll) [reply _posts/2020-12-22-aws-antipatterns.md:28]: Good catch, I'll add a sentence here.

File: _posts/2020-12-22-aws-antipatterns.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (crystoll) [reply _posts/2020-12-22-aws-antipatterns.md:42]: Yes

File: _posts/2020-12-22-aws-antipatterns.md (hunk 8 above)
Comment (crystoll) [reply _posts/2020-12-22-aws-antipatterns.md:166]: Ah, good find! This is not finished, I will fix this,.

**Review by denzilferreira — APPROVED** [review]
Looks good!

**Review by LassiAutioSolita — APPROVED** [review]

**Pull request merged**